Library/Homebrew/tap_auditor.rb

```diff
@@ -4,59 +4,67 @@
 module Homebrew
   # Auditor for checking common violations in {Tap}s.
   class TapAuditor
-    attr_reader :name, :path, :formula_names, :formula_aliases, :formula_renames, :cask_tokens,
-                :tap_audit_exceptions, :tap_style_exceptions, :tap_pypi_formula_mappings, :problems
+    sig { returns(Tap) }
+    attr_reader :tap
+
+    attr_reader :problems
 
     sig { params(tap: Tap, strict: T.nilable(T::Boolean)).void }
-    def initialize(tap, strict:)
-      Homebrew.with_no_api_env do
-        @name                      = tap.name
-        @path                      = tap.path
-        @tap_audit_exceptions      = tap.audit_exceptions
-        @tap_style_exceptions      = tap.style_exceptions
-        @tap_pypi_formula_mappings = tap.pypi_formula_mappings
-        @problems                  = []
-
-        @cask_tokens = tap.cask_tokens.map do |cask_token|
-          cask_token.split("/").last
-        end
-        @formula_aliases = tap.aliases.map do |formula_alias|
-          formula_alias.split("/").last
-        end
-        @formula_renames = tap.formula_renames
-        @formula_names = tap.formula_names.map do |formula_name|
-          formula_name.split("/").last
-        end
-      end
+    def initialize(tap, strict: nil)
+      @tap = tap
+      @problems = []
     end
 
     sig { void }
     def audit
-      audit_json_files
-      audit_tap_formula_lists
-      audit_aliases_renames_duplicates
+      Homebrew.with_no_api_env do
+        audit_json_files
+        audit_ruby_files
+        audit_tap_formula_lists
+        audit_aliases_renames_duplicates
+      end
     end
 
     sig { void }
     def audit_json_files
-      json_patterns = Tap::HOMEBREW_TAP_JSON_FILES.map { |pattern| @path/pattern }
+      json_patterns = Tap::HOMEBREW_TAP_JSON_FILES.map { |pattern| tap.path/pattern }
       Pathname.glob(json_patterns).each do |file|
         JSON.parse file.read
       rescue JSON::ParserError
-        problem "#{file.to_s.delete_prefix("#{@path}/")} contains invalid JSON"
+        problem "#{file.to_s.delete_prefix("#{tap.path}/")} contains invalid JSON"
       end
     end
 
+    sig { void }
+    def audit_ruby_files
+      stray_ruby_files = tap.path.glob("**/*.rb", File::FNM_DOTMATCH)
+
+      stray_ruby_files -= tap.command_files
+      stray_ruby_files.reject! { _1.ascend.any?(tap.command_dir) }
+
+      # Allow mixed formula/cask taps except for core taps.
+      stray_ruby_files -= tap.cask_files unless tap.core_tap?
+      stray_ruby_files -= tap.formula_files unless tap.core_cask_tap?
+
+      return if stray_ruby_files.none?
+
+      problem "Ruby files in wrong location:\n#{stray_ruby_files.map(&:to_s).join("\n")}"
```

Comment:
This should say what location they should be in. Even from reading the code: I don't really understand what would be the right/wrong location here. Can you elaborate?

Reply:
Correct would be `*.rb` (if `Formula/` doesn't exist), `cmd/**/*.rb`, `Formula/**/*.rb` and `Casks/**/*.rb`, anything else incorrect.

Additionally, `Formula/**/*.rb` is incorrect in `homebrew/cask` and `cask/**/*.rb` is incorrect in `homebrew/core`.

Reply:
Homebrew/test-bot is a tap with formulae but also uses `lib/**/*.rb`.

Homebrew/portable-ruby has `Abstract/**.rb`.

Stray files detection here seems somewhat far reaching.

Reply:
> Additionally, `Formula/**/*.rb` is incorrect in `homebrew/cask` and `cask/**/*.rb` is incorrect in `homebrew/core`.

This seems reasonable and a good add. Could also look for Ruby files that look like formulae or casks that aren't in a directory that Homebrew will ever find

> Correct would be `*.rb` (if `Formula/` doesn't exist), `cmd/**/*.rb`, `Formula/**/*.rb` and `Casks/**/*.rb`, anything else incorrect.

Agreed with @Bo98: this does not and feels like it's going to blow up on a _lot_ of taps, official and otherwise.

What's the motivation here?

Reply:
The motivation is to move the logic from the cask CI workflow into `brew audit`. I'm also fine with just removing this check altogether.

Reply:
Which part of the cask CI currently checks this?

Reply:
See https://github.com/Homebrew/homebrew-cask/pull/177890.

Reply:
Personally I think either:
- this check can be removed entirely
- if we want it to be in Homebrew/brew: it should check for files than e.g. contain `class .+ < Formula` or `cask ".+" do` that aren't in `potential_formula_dirs` https://github.com/Homebrew/brew/blob/9f58612deb3f73ef2a6708de47ef90a5f3298a66/Library/Homebrew/tap.rb#L655-L658 or `cask_dir` https://github.com/Homebrew/brew/blob/9f58612deb3f73ef2a6708de47ef90a5f3298a66/Library/Homebrew/tap.rb#L668-L671 and leave all other random Ruby files alone

Reply:
Matching `class .+ < Formula` would still include the `Abstract/*.rb` mentioned above. Then again, I'm not a fan of subclassing `Formula` that way in the first place.

Reply:
Then: remove it entirely, I think.